### ISSUE TITLE
question about subject list in FastSurfer on multiple subjects

### ISSUE THREAD

**Issue by caotihutang** [Deep-MI/FastSurfer#438]
## Question/Support Request

I got an error output "ERROR: Flag unrecognized" and couldn't figure it out. what should I do ? thanks.

## Screenshots

![image](https://github.com/Deep-MI/FastSurfer/assets/48438844/b4cef6f0-e954-4e83-8fb6-4192b1f7f540)

this is my subject list:
subject_id1=/data/subjectx/001.nii.gz\n
subject_id2=/data/subjectx/002.nii.gz\n
subject_id3=/data/subjectx/003.nii.gz\n
subject_id4=/data/subjectx/004.nii.gz\n
subject_id5=/data/subjectx/005.nii.gz\n
subject_id6=/data/subjectx/006.nii.gz\n
subject_id7=/data/subjectx/007.nii.gz\n
subject_id8=/data/subjectx/008.nii.gz\n
subject_id9=/data/subjectx/011.nii.gz\n
subject_id10=/data/subjectx/012.nii.gz\n
subject_id11=/data/subjectx/013.nii.gz\n
subject_id12=/data/subjectx/014.nii.gz\n

## Environment
 - FastSurfer Version:lastest
 - FreeSurfer Version: ... 
 - OS: Windows 10
 - CPU: ...
 - GPU: ...

docker desktop

### Execution

Run Command:

**Comment by dkuegler** [comment]
Just a question about the subject list.
Did you actually add the "\n" at the end of each line?
I think that was just there to show there is a newline.

Also, can you add `--debug` so we can see some extra output of what it is trying to do?

Thanks!

**Comment by caotihutang** [comment]
thanks. 
I just remove the "\n" at the end of each line. I got the same error message.
this is the debug results. 
![image](https://github.com/Deep-MI/FastSurfer/assets/48438844/e5970441-7c10-465b-b1f8-7c6729261c7f)
![image](https://github.com/Deep-MI/FastSurfer/assets/48438844/1f1519ae-8996-41b7-b36e-472678ab59af)

**Comment by dkuegler** [comment]
I think I found the issue. That would indeed be a bug 

In brun_fastsurfer.sh we pass an empty string as a command if neither --seg_only nor --surf_only is specified. https://github.com/Deep-MI/FastSurfer/blob/ec456aaafb0def21f2245848e46dbae1fdd0c69f/brun_fastsurfer.sh#L412

As a workaround, you can just run the script by adding `--seg_only` first, and then run it again with `--surf_only`.

**Comment by caotihutang** [comment]
thanks

**Comment by gDegano** [comment]
Hi there,

I'm writing to follow up on the issue regarding the processing with brun_fastsurfer on multiple subjects. I tried the temporary solution that @dkuegler proposed: using --seg_only first, and then --surf_only. However, this still takes a bit too much time since I cannot parallelize the process as I read in the description of brun_fastsurfer that --parallel_subjects surf is not compatible with either --seg_only or --surf_only (Am I wrong?).
I also tried the branch (dkuegler:slurm/more-commands) in native (Linux Ubuntu 20) but it still gives flag errors.

Command:
./brun_fastsurfer.sh --subject_list $datadir/subjects_list.txt \
                     --sd $fastsurferdir \
                     --asegdkt_segfile --parallel --parallel_subjects 5 --threads 10 --3T

Error messages: 
DEBUG: subject 2: sanlm_swmsMR10033_CC520607-0003-00001-000192-01=/data/sanlm_swmsMR10033_CC520607-0003-00001-000192-01.nii
sanlm_swmsMR10033_CC520597-0003-00001-000192-01: ERROR: Flag unrecognized.
DEBUG: /home/deganog21/ToolboxesGitHub/ModsFastSurfer/FastSurfer//run_fastsurfer.sh  --sid sanlm_swmsMR10033_CC520607-0003-00001-000192-01 --t1 /data/sanlm_swmsMR10033_CC520607-0003-00001-000192-01.nii --sd /home/deganog21/data/team/deganog21/FastSurferSegmentations/CamCam/FastSurfOut/ --asegdkt_segfile --parallel 5 --threads 10 --3T [&]


Do you have any workaround since I need to parallelize (>700 subjects ) and no slurm. I do have a fairly powerful Linux server with Ubuntu 20 (the one used for the above command) with 200GB of memory so I should be able to run at least 8-10 subjects in parallel (no?)

Thanks for your help and the great work!

**Comment by dkuegler** [comment]
`dkuegler:slurm/more-commands` is not final yet (that is why it is not merged :))

If your server has SLURM running on it, that would be the best option. If not, brun_fastsurfer is better.

Please note, that we are no longer supporting Ubuntu 20, I recommend using a docker or singularity image to not have to upgrade your linux distribution.

If you want to run multiple subjects in parallel using brun with the `--parallel_subjects` option would be best; I would recommend something like `--parallel_subjects surf=12` (this will run segmentation and 12 surfaces pipelines at the same time and should not be affected by the bug that was mentioned above (which was cause for the `--seg_only` and the `--surf_only` recommendation). But I know there was an issue with the parsing of the parallel_subjects options, which should be solved by the PR. But this issue might not affect you...

**Comment by dkuegler** [comment]
you probably also want to manually specify the `--threads` as the default is currently 1 here, which  makes this very slow for segmentation on the cpu.

**Comment by gDegano** [comment]
ok great! Sorry btw... I misread the issue and I was making it way more convoluted than necessary... docker seems to work smoothly (for now!;))

Thanks again!

**Comment by gDegano** [comment]
sorry @dkuegler to bother you again but it seems that I keep having an issue after the segmentation.

Command: 
docker run --gpus all -v $DATA_T1:/data \
                      -v $FLD_Fast_Out:/output \
                      -v $FS_license:/fs_license \
                      --entrypoint "/fastsurfer/brun_fastsurfer.sh" \
                      --rm --user $(id -u):$(id -g) deepmi/fastsurfer \
                      --parallel_subjects surf=10 --fs_license /fs_license/license.txt \
                      --sd /output --subject_list /data/subjects_list.txt \
                      --parallel --fsaparc --3T --threads 25 --asegdkt_segfile 

Error (after correct segmentation):
/fastsurfer/brun_fastsurfer.sh: line 412: prepend: command not found

(also --debug does not seems to work (it gives /fastsurfer/brun_fastsurfer.sh: line 257: printf: --: invalid option
printf: usage: printf [-v var] format [arguments]))

**Comment by dkuegler** [comment]
I think those are indeed bugs that happen in a specific combination of arguments. I don't quite know why the prepend error happens, maybe you can say which version exactly you are using?

Also, I am currently testing all the settings for the merge into dev and the current version in #434 is promising.

**Comment by gDegano** [comment]
sorry you are right. I am using the latest docker version at the moment.
Thanks for working on this, it will be a great game changer for us!:)

**Comment by m-reuter** [comment]
#434 has been merged and I am closing this for now. Feel free to re-open if this is not solved.
